group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,891 @@
+     */
+    private void reassignPartition(TopicIdPartition partition) {
+        // Find the new member with the least assignment size.
+        String newOwner = null;
+        for (String anotherMember: sortedMembersByAssignmentSize) {
+            if (members.get(anotherMember).subscribedTopicIds().contains(partition.topicId())) {
+                newOwner = anotherMember;
+                break;
+            }
+        }
+
+        assert newOwner != null;
+
+        reassignPartition(partition, newOwner);
+    }
+
+    /**
+     * Reassigns the given partition to a new member while considering partition movements and stickiness.
+     * <p>
+     * This method performs the following actions:
+     * <ol>
+     *   <li> Determines the current owner of the partition. </li>
+     *   <li> Identifies the correct partition to move, adhering to stickiness constraints. </li>
+     *   <li> Processes the partition movement to the new member. </li>
+     * </ol>
+     *
+     * @param partition     The {@link TopicIdPartition} to be reassigned.
+     * @param newMember     The Id of the member to which the partition should be reassigned.
+     */
+    private void reassignPartition(TopicIdPartition partition, String newMember) {
+        String member = partitionOwnerInTargetAssignment.get(partition);
+        // Find the correct partition movement considering the stickiness requirement.
+        TopicIdPartition partitionToBeMoved = partitionMovements.getTheActualPartitionToBeMoved(
+            partition,
+            member,
+            newMember
+        );
+        processPartitionMovement(partitionToBeMoved, newMember);
+    }
+
+    private void processPartitionMovement(TopicIdPartition topicIdPartition, String newMember) {
+        String oldMember = partitionOwnerInTargetAssignment.get(topicIdPartition);
+
+        partitionMovements.movePartition(topicIdPartition, oldMember, newMember);
+
+        assignmentManager.removePartitionFromMembersAssignment(topicIdPartition, oldMember);
+        assignmentManager.addPartitionToMembersAssignment(topicIdPartition, newMember);
+    }
+
+    /**
+     * <code> MemberPair </code> represents a pair of member Ids involved in a partition reassignment. Each
+     * <code> MemberPair </code> object, which contains a source (<code>src</code>) and a destination (<code>dst</code>)
+     * element, normally corresponds to a particular partition or topic, and indicates that the particular partition or some
+     * partition of the particular topic was moved from the source member to the destination member during the rebalance.
+     */
+    private static class MemberPair {
+        private final String srcMemberId;
+        private final String dstMemberId;
+
+        MemberPair(String srcMemberId, String dstMemberId) {
+            this.srcMemberId = srcMemberId;
+            this.dstMemberId = dstMemberId;
+        }
+
+        public String toString() {
```

Comment:
nit: Add `@Override`. We usually also put toString last in the class.

Reply:
yes that's an oversight on my part thanks!